### PR TITLE
Issue #2713891 by mglaman: Ability to use fields on an commerce_produ…

### PR DESCRIPTION
**Pull request by mglaman**
…ct_attribute_value for selecting that attribute

Doesn't currently have a test.

### CODE REVIEW

File: modules/product/src/Element/CommerceProductRenderedAttribute.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use Drupal\Component\Utility\Html as HtmlUtility;
+
+/**
+ * Provides a form input element for rendering attributes as radio buttons.
+ *
+ * Example usage:
+ * @code
+ * $form['settings']['active'] = array(
+ *   '#type' => 'commerce_product_rendered_attribute',
+ *   '#title' => $this->t('Poll status'),
```
Comment (bojanz) [inline modules/product/src/Element/CommerceProductRenderedAttribute.php]: Poll status is a bit of a weird example :) Looks like we need to rewrite the entire docblock.

File: modules/product/src/Element/CommerceProductRenderedAttribute.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      /** @var \Drupal\Core\Entity\EntityStorageInterface $storage */
+      $storage = \Drupal::service('entity_type.manager')->getStorage('commerce_product_attribute_value');
+      $attribute_values = $storage->loadMultiple(array_keys($element['#options']));
+
+      /** @var \Drupal\Core\Render\RendererInterface $renderer */
+      $renderer = \Drupal::service('renderer');
+
+      $weight = 0;
+      foreach ($element['#options'] as $key => $choice) {
+        $rendered_attribute = \Drupal::entityTypeManager()->getViewBuilder('commerce_product_attribute_value')->view($attribute_values[$key], 'add_to_cart');
```
Comment (bojanz) [inline modules/product/src/Element/CommerceProductRenderedAttribute.php]: Let's move $view_builder =  \Drupal::entityTypeManager()->getViewBuilder('commerce_product_attribute_value') to its own line, to shorten this.

File: modules/product/src/Entity/ProductAttribute.php [hunk 3]
```diff
@@ -70,6 +78,13 @@ public function getValues() {
   /**
    * {@inheritdoc}
    */
+  public function getElementType() {
+    return (!$this->elementType) ? 'select' : $this->elementType;
```
Comment (bojanz) [inline modules/product/src/Entity/ProductAttribute.php]: Can't the default be provided on the class attribute?

File: modules/product/src/Entity/ProductAttributeInterface.php [hunk 4]
```diff
@@ -16,4 +16,10 @@
    */
   public function getValues();
 
+  /**
+   * Gets the attribute element type.
+   * 
+   * @return string
+   */
+  public function getElementType();
```
Comment (bojanz) [inline modules/product/src/Entity/ProductAttributeInterface.php:25]: Missing empty line before }

File: modules/product/src/ProductAttributeFieldManager.php [hunk 5]
```diff
@@ -143,7 +143,7 @@ protected function buildFieldMap() {
         $handler_settings = $definition->getSetting('handler_settings');
         $component = $form_display->getComponent($field_name);
 
-        $field_map[$bundle][] = [
+        $field_map[$bundle][$field_name] = [
```
Comment (bojanz) [inline modules/product/src/ProductAttributeFieldManager.php]: Why are we making this change?

**Comment by bojanz** [comment]
We need to remove all mentions of attribute_widget_type (the widget setting).

We also need the JS for the element: http://cgit.drupalcode.org/commerce_fancy_attributes/tree/commerce_fancy_attributes.js

That, the CSS, the form element might still belong in commerce_fancy_attributes, since the user will need that module anyway (for the color swatch formatter).

File: modules/product/src/ProductAttributeFieldManager.php (hunk 5 above)
Comment (mglaman) [inline modules/product/src/ProductAttributeFieldManager.php]: See https://github.com/drupalcommerce/commerce/pull/368/files#diff-354d71e20ba2cfced9d85f311aefa77cR303

So we can foreach, use key for field name and still have attribute name for that field (so we can load it, see how it should render.) I feel there is a better approach, but last night it was +1 to get it done.

**Comment by mglaman** [comment]
@bojanz why do we need the JS? the CSS is hiding the radio button and it is being handled as is? I figured commerce_fancy_attributes would be our placeholder for crazy field formatters (like hex code to square swatch.)

**Comment by mglaman** [comment]
@bojanz we have passing tests :)

File: modules/product/src/Form/ProductAttributeForm.php [hunk 6]
```diff
@@ -3,13 +3,44 @@
...
+  protected $entityFieldManager;
+
+  /**
+   * Constructs a new ProductAttributeForm object.
+   *
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   The entity field manager.
+   */
+  public function __construct(EntityFieldManagerInterface $entity_field_manager) {
+    $this->entityFieldManager = $entity_field_manager;
```
Comment (bojanz) [inline modules/product/src/Form/ProductAttributeForm.php]: You don't appear to be using the entity field manager anywhere?

File: modules/product/src/Plugin/Field/FieldWidget/ProductVariationAttributesWidget.php [hunk 7]
```diff
@@ -96,57 +104,6 @@ public static function isApplicable(FieldDefinitionInterface $field_definition)
   /**
    * {@inheritdoc}
    */
-  public static function defaultSettings() {
-    return [
-      'attribute_widget_type' => 'select',
```
Comment (bojanz) [inline modules/product/src/Plugin/Field/FieldWidget/ProductVariationAttributesWidget.php]: There's also the schema in attribute_widget_type.
We also need to re-export core.entity_form_display.commerce_line_item.product_variation.add_to_cart.yml to remove the now-gone setting.

File: modules/product/src/Tests/ProductRenderedAttributeTest.php [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  protected function setUp() {
+    parent::setUp();
+
+    $this->attributeFieldManager = \Drupal::service('commerce_product.attribute_field_manager');
+  }
+
+  /**
+   * Tests that the add to cart form renders an attribute entity.
+   */
+  public function testRenderedAttributeElement() {
```
Comment (bojanz) [inline modules/product/src/Tests/ProductRenderedAttributeTest.php]: Why not put this method in the old AddToCartFormTest?

Comment (mglaman) [inline modules/product/src/Tests/ProductRenderedAttributeTest.php]: Good call. I'm actually having to borrow two methods from it anyways.

File: modules/product/src/Plugin/Field/FieldWidget/ProductVariationAttributesWidget.php (hunk 7 above)
Comment (mglaman) [inline modules/product/src/Plugin/Field/FieldWidget/ProductVariationAttributesWidget.php]: Done!

File: modules/cart/src/Tests/AddToCartFormTest.php [hunk 9]
```diff
@@ -270,6 +271,60 @@ public function testMultipleVariations() {
...
+
+    $this->drupalGet($product->toUrl());
+    $this->assertOptionWithDrupalSelector('edit-purchased-entity-0-attributes-attribute-color', $color_attribute_values['cyan']->id());
+
+    $color_attribute->set('elementType', 'commerce_product_rendered_attribute')->save();
+
+    $this->drupalGet($product->toUrl());
+    $this->assertText('Cyan (Rendered)');
+    $this->assertText('Magenta (Rendered)');
+
```
Comment (bojanz) [inline modules/cart/src/Tests/AddToCartFormTest.php]: Extra newline.

File: modules/product/commerce_product.libraries.yml [hunk 10]
```diff
@@ -3,3 +3,8 @@ form:
   css:
     theme:
       css/product.form.css: {}
+rendered-attributes:
```
Comment (bojanz) [inline modules/product/commerce_product.libraries.yml:7]: Let's add a newline above rendered-attributes.

File: modules/product/css/product.rendered-attributes.css [hunk 11]
```diff
@@ -0,0 +1,14 @@
+.product--rendered-attribute .form-item {
```
Comment (bojanz) [inline modules/product/css/product.rendered-attributes.css]: This file should be commerce_product.rendered-attributes.css and have a docblock on top.

File: modules/product/src/Entity/ProductAttributeInterface.php [hunk 12]
```diff
@@ -16,4 +16,11 @@
    */
   public function getValues();
 
+  /**
+   * Gets the attribute element type.
+   *
+   * @return string
```
Comment (bojanz) [inline modules/product/src/Entity/ProductAttributeInterface.php:22]: Missing description below @return string

Comment (mglaman) [inline modules/product/src/Entity/ProductAttributeInterface.php:22]: `The element type name.`

File: modules/product/src/Element/CommerceProductRenderedAttribute.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * The options must be an array of attribute values, keyed by the entity's ID.
+ *
+ * Example usage:
+ * @code
+ * $form['rendered_attributes'] = array(
+ *   '#type' => 'commerce_product_rendered_attribute',
+ *   '#title' => $this->t('Attributes'),
+ *   '#default_value' => 1,
+ *   '#options' => array(0 => 'Red', 1 => 'Blue),
```
Comment (bojanz) [inline modules/product/src/Element/CommerceProductRenderedAttribute.php]: Let's not use array() even in examples.

**Pull request merged**